aiokafka/client.py

```diff
@@ -1,489 +1,335 @@
-        but has no leader.
-
-        UnknownTopicOrPartitionError will be raised if the topic or partition
-        is not part of the metadata.
-
-        LeaderNotAvailableError is raised if server has metadata, but there is
-        no current leader
-        """
-
-        key = TopicAndPartition(topic, partition)
-
-        # Use cached metadata if it is there
-        if self._topics_to_brokers.get(key) is not None:
-            return self._topics_to_brokers[key]
-
-        # Otherwise refresh metadata
-
-        # If topic does not already exist, this will raise
-        # UnknownTopicOrPartitionError if not auto-creating
-        # LeaderNotAvailableError otherwise until partitions are created
-        yield from self.load_metadata_for_topics(topic)
-
-        # If the partition doesn't actually exist, raise
-        if partition not in self._topic_partitions[topic]:
-            raise UnknownTopicOrPartitionError(key)
-
-        # If there's no leader for the partition, raise
-        meta = self._topic_partitions[topic][partition]
-        if meta.leader == -1:
-            raise LeaderNotAvailableError(meta)
-
-        # Otherwise return the BrokerMetadata
-        return self._brokers[meta.leader]
-
-    def _next_id(self):
-        """
-        Generate a new correlation id
-        """
-        self._request_id += 1
-        if self._request_id > 0x7fffffff:
-            self._request_id = 0
-        return self._request_id
-
-    @asyncio.coroutine
-    def _send_broker_unaware_request(self, payloads, encoder_fn, decoder_fn):
-        """
-        Attempt to send a broker-agnostic request to one of the available
-        brokers. Keep trying until you succeed.
-        """
-        for (host, port) in self._hosts:
-            request_id = self._next_id()
+    def close(self):
+        if self._sync_task:
+            self._sync_task.cancel()
             try:
-                conn = yield from self._get_conn(host, port)
-                request = encoder_fn(client_id=self._client_id,
-                                     correlation_id=request_id,
-                                     payloads=payloads)
-
-                try:
-                    fut = conn.send(request)
-                    response = yield from fut
-                except Exception:
-                    self._drop_conn(conn)
-                    raise
-                return decoder_fn(response)
-
-            except Exception as e:
-                log.warning("Could not send request [%r] to server %s:%i, "
-                            "trying next server: %s",
-                            request_id, host, port, e)
-
-        raise KafkaUnavailableError("All servers failed to process request")
+                yield from self._sync_task
+            except asyncio.CancelledError:
+                pass
+            self._sync_task = None
+        for conn in self._conns.values():
+            conn.close()
```

Comment:
Should this method be a function? It does cancel coroutines inside of it, and we can wait for it to finish.